stream_alert/apps/_apps/gsuite.py

```diff
@@ -151,11 +161,24 @@ def _gather_logs(self):
         # once during the first poll
         if not self._next_page_token:
             self._last_timestamp = activities[0]['id']['time']
-            LOGGER.debug('Caching last timestamp: %s', self._last_timestamp)
+            LOGGER.debug('[%s] Caching last timestamp: %s', self, self._last_timestamp)
+            # Store the event ids with the most recent timestamp to de-duplicate them next time
+            next_event_ids = [
+                activity['id']['uniqueQualifier']
+                for activity in activities
+                if activity['id']['time'] == self._last_timestamp
+            ]
+            self._context['next_event_ids'] = next_event_ids[:self._MAX_EVENT_IDS]
+            if len(next_event_ids) > self._MAX_EVENT_IDS:
+                LOGGER.warning('[%s] More than %d next_event_ids. Unable to de-duplicate: %s',
+                               self, self._MAX_EVENT_IDS, next_event_ids[self._MAX_EVENT_IDS:])
 
         self._next_page_token = results.get('nextPageToken')
         self._more_to_poll = bool(self._next_page_token)
 
+        if not self._more_to_poll:
+            self._context['last_event_ids'] = self._context.get('next_event_ids', [])
```

Comment:
this could then just save the cached instance property instead of copying a value that already exist in the context

Reply:
It's true, at this point we no longer need to store `next_event_ids`. But only because there is nothing left to poll, and we know the next lambda invocation will be for a new time period. We could clear `next_event_ids`, but the important thing is to copy them across to `last_event_ids`.

Reply:
Refactored to only store one copy in the context since `_gather_logs` is called multiple times per execution. The case where it gets deferred to the next lambda execution is a very small edge case, and worst case you get some duplicate events.